lib/Dialect/HW/Transforms/HWVectorization.cpp

```diff
@@ -0,0 +1,879 @@
+    }
+    Value replicatedSel =
+        comb::ReplicateOp::create(builder, loc, resultType, sel);
+    vectorizedResult =
+        comb::MuxOp::create(builder, loc, replicatedSel, vectorizedOperands[1],
+                            vectorizedOperands[2]);
+  } else {
+    return nullptr;
+  }
+
+  vectorizedMap[slice0Val] = vectorizedResult;
+  return vectorizedResult;
+}
+
+mlir::Value vectorizer::findBitSource(mlir::Value vectorVal, unsigned bitIndex,
+                                      int depth) {
+  if (auto blockArg = dyn_cast<BlockArgument>(vectorVal)) {
+    if (blockArg.getType().isInteger(1)) {
+      return blockArg;
+    }
+    return nullptr;
+  }
+
+  Operation *op = vectorVal.getDefiningOp();
+  if (!op) {
+    return nullptr;
+  }
+
+  if (op->getNumResults() == 1 && op->getResult(0).getType().isInteger(1)) {
+    return op->getResult(0);
+  }
+
+  if (auto constOp = dyn_cast<hw::ConstantOp>(op)) {
+    if (constOp.getType().isInteger(1)) {
+      return constOp.getResult();
+    }
+    return nullptr;
+  }
+
+  if (auto concat = dyn_cast<comb::ConcatOp>(op)) {
+    unsigned currentBit = cast<IntegerType>(vectorVal.getType()).getWidth();
+    for (Value operand : concat.getInputs()) {
+      unsigned operandWidth = cast<IntegerType>(operand.getType()).getWidth();
+      currentBit -= operandWidth;
+      if (bitIndex >= currentBit && bitIndex < currentBit + operandWidth) {
+        return findBitSource(operand, bitIndex - currentBit, depth + 1);
+      }
+    }
+  } else if (auto orOp = dyn_cast<comb::OrOp>(op)) {
+    if (auto source = findBitSource(orOp.getInputs()[1], bitIndex, depth + 1)) {
+      if (auto sourceConst =
+              dyn_cast_or_null<hw::ConstantOp>(source.getDefiningOp())) {
+        if (!sourceConst.getValue().isZero())
+          return source;
+      } else {
+        return source;
+      }
+    }
+    return findBitSource(orOp.getInputs()[0], bitIndex, depth + 1);
+  } else if (auto andOp = dyn_cast<comb::AndOp>(op)) {
+    Value lhs = andOp.getInputs()[0];
+    Value rhs = andOp.getInputs()[1];
+    if (isa_and_nonnull<hw::ConstantOp>(rhs.getDefiningOp()))
+      return findBitSource(lhs, bitIndex, depth + 1);
+    if (isa_and_nonnull<hw::ConstantOp>(lhs.getDefiningOp()))
+      return findBitSource(rhs, bitIndex, depth + 1);
+  }
+
+  return nullptr;
+}
+
+bool vectorizer::cleanup_dead_ops(Block &block) {
+  bool overallChanged = false;
+  bool changedInIteration = true;
+  while (changedInIteration) {
+    changedInIteration = false;
+    llvm::SmallVector<Operation *, 16> deadOps;
+    for (Operation &op : block) {
+      if (op.use_empty() && !op.hasTrait<mlir::OpTrait::IsTerminator>()) {
+        deadOps.push_back(&op);
+      }
+    }
+    if (!deadOps.empty()) {
+      changedInIteration = true;
+      overallChanged = true;
+      for (Operation *op : deadOps) {
+        op->erase();
+      }
+    }
+  }
+  return overallChanged;
+}
```

Comment:
Could you use https://github.com/llvm/llvm-project/blob/db557bee1e2c128e77805deb86c1f364b5c29e70/mlir/lib/Transforms/Utils/RegionUtils.cpp#L495? There are few issues around side-effecting op and O(N^2) fixpoint iterations here so would be nice to simply use a library function.